sw/host/opentitanlib/src/util/tty.rs

```diff
@@ -0,0 +1,155 @@
+    pub vid: u16,                      // USB vendor ID.
+    pub pid: u16,                      // USB vendor ID.
+    pub serial_number: Option<String>, // USB serial number (if any).
+    pub interface: u8,                 // USB Interface number.
+    pub config: u8,                    // USB configuration number.
+    pub bus: u8,                       // USB bus number.
+    pub ports: Vec<u8>,                // USB port numbers.
+}
+
+/// Path to the linux sysfs USB device tree.
+pub const SYSFS_USB_PATH: &str = "/sys/bus/usb/devices";
+
+pub fn scan_usb_ttys() -> Result<Vec<UsbTty>> {
+    // The devices directory looks something like this (see https://www.kernel.org/doc/Documentation/ABI/stable/sysfs-bus-usb):
+    //
+    // ```plain
+    // 3-6/         # USB device
+    // |- 3-6:1.0/  # USB Interface
+    // |- 3-6:1.1/  # USB Interface
+    // |- 3-6:1.2/  # USB Interface
+    // |- 3-6:1.3/  # USB Interface
+    // |- 3-6:1.3/  # USB Interface
+    // |- power/    # Other device stuff...
+    // |- idVendor  # USB vendor ID
+    // |- idProduct # USB product ID
+    // |- ...
+    //
+    // 4-2/         # USB device
+    // |- 4-2:1.0/  # USB Interface
+    // |- ...
+    // ...
+    // ```
+    //
+    // Note that devices under several hubs have a more complex path like 3-1.4.3 which is more generally
+    // of the form <bus>-<port1>[.<portX>]*
+    // And the interface are numbered as <config>.<interface>
+    let mut list = Vec::new();
+
+    // Regex for a full usb path: <bus>-<port1>[.<portX>]*:<cfg>.<intf>
+    let usb_re = Regex::new(&format!(
+        r"^(?<bus>[0-9]+)-(?<ports>[0-9]+(?:\.[0-9]+)*):(?<cfg>[0-9]+)\.(?<intf>[0-9]+)$"
+    ))
+    .context("Could not build regex to match interfaces")?;
+
+    for dev in fs::read_dir(SYSFS_USB_PATH).context("Cannot find USB device in sysfs")? {
+        let dirent = dev?;
+        let dev_path = dirent.path();
+        let dev_name = dirent
+            .file_name()
+            .to_str()
+            .context("Cannot convert filename to string")?
+            .to_string();
+        // Only consider devices.
+        if !dev_path.is_dir() || dev_name.contains(":") {
+            continue;
+        }
+        // Read VID, PID and serial.
+        let vid = u16::from_str_radix(&fs::read_to_string(dev_path.join("idVendor"))?.trim(), 16)?;
+        let pid = u16::from_str_radix(&fs::read_to_string(dev_path.join("idProduct"))?.trim(), 16)?;
+        let serial_number_path = dev_path.join("serial");
+        let serial_number = if serial_number_path.exists() {
+            Some(fs::read_to_string(serial_number_path)?.trim().to_string())
+        } else {
+            None
+        };
+
+        // List interfaces.
+        for iface in dev_path.read_dir()? {
+            let iface_dirent = iface?;
+            let iface_name = iface_dirent
+                .file_name()
+                .to_str()
+                .context("bad")?
```

Comment:
Batter error message? Also apply on line 102

Reply:
yes this is just for debugging, it's not the final version